secant/Features/BackupFlow/DropDelegate.swift

```diff
@@ -0,0 +1,81 @@
+//
+//  DropDelegate.swift
+//  secant-testnet
+//
+//  Created by Francisco Gindre on 11/16/21.
+//
+
+import Foundation
+import SwiftUI
+import OrderedCollections
+import ComposableArchitecture
+
+/// There's no way to pass a nullable action to a droppable target. So, the Null Object Pattern comes to the rescue
+struct NullDelegate: DropDelegate {
+    func validateDrop(info: DropInfo) -> Bool {
+        return false
+    }
+
+    func performDrop(info: DropInfo) -> Bool {
+        false
+    }
+}
+
+/// Drop delegate that accepts items conforming to `PhraseChip.validationWordTypeIdentifier`
+struct WordChipDropDelegate: DropDelegate {
+    var dropAction: ((PhraseChip.Kind) -> Void)?
+
+    func validateDrop(info: DropInfo) -> Bool {
+        return  info.hasItemsConforming(to: [PhraseChip.validationWordTypeIdentifier])
+    }
+
+    func performDrop(info: DropInfo) -> Bool {
+        if let item = info.itemProviders(for: [PhraseChip.validationWordTypeIdentifier]).first {
+            item.loadItem(forTypeIdentifier: PhraseChip.validationWordTypeIdentifier, options: nil) { text, _ in
+                DispatchQueue.main.async {
+                    if let data = text as? Data {
+                        //  Extract string from data
+
+                        let word = String(decoding: data, as: UTF8.self)
+                        dropAction?(.unassigned(word: word as String))
+                    }
+                }
+            }
+            return true
+        }
+        return false
+    }
+}
+
+extension RecoveryPhraseValidationState {
+    func dropDelegate(
+        for viewStore: RecoveryPhraseValidationViewStore,
+        group: Int
+    ) -> DropDelegate {
```

Comment:
I'm trying to understand the reasoning behind the logic in this function.  If the user in the in the `initial` state, how can they _not_ be in the state where the chipKind is unassigned?


When the user is in the `incomplete` state, and a word is grabbed from a group but the group doesn't have a matching index, is that not an assertion failure type situation, where that state should technically never happen?  Do we allow users to put a word back or once they have chosen are they stuck with that position?

Once we are in the `complete` phase, can a user even do drag and drop? Do we allow them to put words back?

What I'm wondering is, given these questions about how this interaction works, do we even need a delegate that always returns false?  It seems like the situations where we are using one are largely error states or simply not possible to get into.

Reply:
I've double-checked and the current logic no longer requires to pass a NullDelegate.

Reply:
That's great news! That's huge!